HTTPS Sender

iori-yja opened the pull request:
<!--
We appreciate your contribution to the Jaeger project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- You have read the guide for contributing
  - See https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md#sign-your-work
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Resolves #123"
-->

## Which problem is this PR solving?
- Allow to send to collector via https, even when the collector's https sidecar is self-signed, key-continuity-based backend: https://github.com/jaegertracing/jaeger-client-java/issues/595
- Related: https://github.com/jaegertracing/jaeger/issues/458

## Short description of the changes
Added those below:

- Optional Certificate Pinning
- Optional Certificate verification disabling (used only with pinning)
- Integration test over HTTPS

jpkrohling commented:
Looks like it failed the checkstyle:

```
[ant:checkstyle] [ERROR] /home/travis/build/jaegertracing/jaeger-client-java/jaeger-crossdock/src/main/java/io/jaegertracing/crossdock/JerseyServer.java:147: Line is longer than 120 characters (found 140). [LineLength]
```

iori-yja commented:
Oh, that was too stupid. I believe it's fixed now.
Also, crossdock check is still broken. It turns out that the cause is Bad Gateway on call to jaeger-collector from its https-proxy, but it seems weird...

yurishkuro commented:
> Also, crossdock check is still broken. It turns out that the cause is Bad Gateway on call to jaeger-collector from its https-proxy, but it seems weird...

if you found this in the build logs, could you copy them here?  Many PRs have been seeing crossdock failures recently, I suspect a common cause.

iori-yja commented:
I've added an nginx proxy in the docker-compose.yml. This service's log shows `no route to host`(pasted below).
For debug purpose, I added host-port mapping (14443:14443) as in the current diff and saw Bad Gateway response from `localhost:14443` with my browser (chromium, curl) on the host machine.

```
% docker-compose -f jaeger-crossdock/docker-compose.yml -f jaeger-crossdock/jaeger-docker-compose.yml logs jaeger-collector-https-proxy
Attaching to jaeger-crossdock_jaeger-collector-https-proxy_1
jaeger-collector-https-proxy_1  | 2019/03/10 11:55:40 [error] 6#6: *1 connect() failed (113: No route to host) while connecting to upstream, client: 192.168.64.8, server: jaeger-collector-https-proxy, request: "POST /api/traces?format=jaeger.thrift HTTP/1.1", upstream: "http://192.168.64.9:14268/api/traces?format=jaeger.thrift", host: "jaeger-collector-https-proxy:14443"
jaeger-collector-https-proxy_1  | 192.168.64.8 - - [10/Mar/2019:11:55:40 +0000] "POST /api/traces?format=jaeger.thrift HTTP/1.1" 502 157 "-" "okhttp/3.9.0" "-"
```


I don't think the network problem is caused by reporter misconfiguration (as I changed it) because it succeeded to establish a session to the proxy, and also I don't think collector's issue because plain http version (java-http) is able to report to it.

This failure happens very likely but merely it succeeds, if I remember correctly (I couldn't reproduce now).

iori-yja commented:
I tried once again which was successful this time (no change).
```
% make crossdock-clean && make
(snip)
2019/03/10 15:57:22 HTTP: HEAD http://test_driver:8080/
HTTP: Service test_driver:8080 is up

All services are up after 15.10839437s!

Executing Matrix...

...................................................................................................................................

131/131 passed (0/131 skipped)

Tests passed!

% git diff
% docker-compose -f jaeger-crossdock/docker-compose.yml -f jaeger-crossdock/jaeger-docker-compose.yml logs jaeger-collector-https-proxy
Attaching to jaeger-crossdock_jaeger-collector-https-proxy_1
jaeger-collector-https-proxy_1  | 192.168.80.8 - - [10/Mar/2019:15:57:25 +0000] "POST /api/traces?format=jaeger.thrift HTTP/1.1" 202 0 "-" "okhttp/3.9.0" "-"
```

yurishkuro commented on jaeger-core/src/main/java/io/jaegertracing/Configuration.java:
```diff
@@ -152,6 +152,12 @@
+   *  TLS certificates (in comma-separated BASE64 SHA256 Hash) for certificates pinning,
+   *  used in case of HTTPS communication to the endpoint.
+   */
+  public static final String JAEGER_TLS_CERTIFICATE_PINNING = JAEGER_PREFIX + "TLS_CERTIFICATE_PINNING";
```
Not thrilled with the name. Is it typical to pass certs via env vars?  We've recently added many TLS options to Jaeger backend, and the certs are always passed via files.

iori-yja replied:
I had three reasons to pass this via environment variable over file.

1. It should be good to be consistent with other reporter configurations most of them can be done via fromEnv() otherwise manipulated with Internal*.
1. This must be SHA256 Hash of Certificate or Public Key (https://square.github.io/okhttp/3.x/okhttp/okhttp3/CertificatePinner.html) so this is kept short.
1. In Lambda, one of the motivations for this, managing environment variables is easier than doing so for file. Environment variables are shown and settable plainly in the console, while files are to be uploaded directly or indirectly (S3) to the console as zipped format.

iori-yja commented:
Unchanged files drop coverage :thinking:  https://codecov.io/gh/jaegertracing/jaeger-client-java/pull/602/changes

Retrying by merge the remote HEAD.

yurishkuro commented on jaeger-crossdock/https-proxy/Dockerfile:
```diff
@@ -0,0 +1,8 @@
+ADD build/proxy.key /etc/nginx/proxy.key
+
+EXPOSE 8080 14443
+CMD ["sh", "-c", "while ! { wget --spider -S http://jaeger-collector:14269; }; do echo waiting...; sleep 1; done; nginx -g 'daemon off;'"]
```
comment would be useful, what is being queried on port 14269?

iori-yja replied:
added.

yurishkuro commented on jaeger-crossdock/rules.mk:
```diff
@@ -5,14 +5,14 @@ JAEGER_COMPOSE_URL=https://raw.githubusercontent.com/jaegertracing/jaeger/master
+crossdock: gradle-compile crossdock-proxy-secret-gen crossdock-download-jaeger
+	docker-compose -f $(XDOCK_YAML) -f $(XDOCK_JAEGER_YAML) kill java-udp java-http java-https
+	docker-compose -f $(XDOCK_YAML) -f $(XDOCK_JAEGER_YAML) rm -f java-udp java-http java-https
+	docker-compose -f $(XDOCK_YAML) -f $(XDOCK_JAEGER_YAML) build java-udp java-http java-https
```
* should this also include jaeger-collector-https-proxy ?
* I would move these into a make var, to DRY

yurishkuro commented on jaeger-crossdock/https-proxy/authority.key:
```diff
@@ -0,0 +1,8 @@
+MHcCAQEEIAA24MB8sxrLG1an0nG1DCH6J32iqrtborxFOjdqWNCmoAoGCCqGSM49
+AwEHoUQDQgAEAhYnw9zYU0G3VZ48nNlT5jAs096pX0zeHM/yxiJe+DS5Yj0EJXM/
+0A1Of2zqxbyJpaEIFcqTmTTyTXCE7I6B6Q==
+-----END EC PRIVATE KEY-----
```
are these being generated by `jaeger-crossdock/https-proxy/gen.sh` every time?

iori-yja replied:
No, this is root certificate which is used to generate proxy's certificate by `gen.sh`. Hard-coding is ugly by the way, but sometimes test project includes.

The Public key pin is coded here:
https://github.com/jaegertracing/jaeger-client-java/pull/602/files#diff-9d280e9b01ffb0fc60647fe5d2a8286cR74

iori-yja replied:
OK, I will.

yurishkuro commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -66,6 +77,7 @@ public void send(Process process, List<Span> spans) throws SenderException {
     try {
       response = httpClient.newCall(request).execute();
     } catch (IOException e) {
+      e.printStackTrace();
```
don't we have that elsewhere?

iori-yja replied:
That was my mistake not to remove debug code. I will delete.

yurishkuro commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -119,13 +135,96 @@ public Builder withAuth(String authToken) {
+    }
+
+    public Builder acceptSelfSigned() {
+      // This dangerous operation will only take effect if pinning is used.
```
method comment?

iori-yja replied:
Not intended but should be.

yurishkuro commented:
I feel like this is adding way too much code to the client.

iori-yja commented on jaeger-crossdock/docker-compose.yml:
```diff
@@ -64,11 +65,36 @@ services:
+            - jaeger-collector
+
+    jaeger-collector-https-proxy:
+        image: nginx:alpine
```
This is nice to have for jaeger, and I may add the option to jaeger soon, but this proxy is for client test for now. I will replace this after `--collector.http.tls` options are merged.

jpkrohling commented on jaeger-crossdock/https-proxy/proxy.template.conf:
```diff
@@ -0,0 +1,20 @@
+    listen 14443 ssl;
+    server_name ${SERVER};
+
+    ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
```
I know this is only for testing, but could we have a production-grade configuration here? Like, a secure set of ciphers and options like `ssl_prefer_server_ciphers on`. This way, we'd be testing with production settings, making sure that all involved components are up to modern standards.

For reference, this is what I used to have on a few servers of mine, which was "current" a couple of years ago:
```nginx
server {
	listen [::]:443 ssl http2;
	listen 443 ssl http2;
	server_name acme.example.com;
	ssl_certificate /etc/pki/tls/certs/acme.example.com.bundle.cert;
	ssl_certificate_key /etc/pki/tls/private/acme.example.com.key;
	ssl_session_cache shared:SSL:10m;
	ssl_session_timeout 10m;
	ssl_protocols TLSv1 TLSv1.1 TLSv1.2; # probably needs an update to include TLSv1.3?
	ssl_prefer_server_ciphers on;
	ssl_ciphers "EECDH+ECDSA+AESGCM EECDH+aRSA+AESGCM EECDH+ECDSA+SHA384 EECDH+ECDSA+SHA256 EECDH+aRSA+SHA384 EECDH+aRSA+SHA256 EECDH !EDH+aRSA !RC4 !aNULL !eNULL !LOW !3DES !MD5 !EXP !PSK !SRP !DSS";
	add_header Strict-Transport-Security max-age=31536000;
	root /usr/share/nginx/html/acme.example.com;
}

```

jpkrohling commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -119,13 +134,98 @@ public Builder withAuth(String authToken) {
+           * Instead, check the sha256 hash value by custom verifier. */
+          acceptSelfSigned(clientBuilder, hostname, pins);
+        }
+      } catch (Exception e) {
```
This deserves a log message, even if at `debug` level (but possibly at info or even warn, as this is called only once)

iori-yja replied:
I afraid it isn't, because it is very similar to https://github.com/jaegertracing/jaeger-client-java/pull/602/files/a0607f48cd7370959b8f937a208fdc517a03b82f#diff-3330f8a6420c943a90a1b22b0abd0815R56. But it might be OK to fail earlier than build itself.

jpkrohling replied:
Not sure I understood what you mean. The code you linked throw an exception, the code here is swallowing one. Instead of swallowing, we should log it at some level (`info` in my opinion).

iori-yja replied:
The `try` above the `catch` is needed because of `URI` to be constructed by the `endpoint` which is supplied by the configuration. If the endpoint is invalid, building `HttpSender` at few lines later would fail. But I agree with that it shouldn't swallow all kind of `Exception`, so I changed it to catch only `URISyntaxException` and throw it earlier.

jpkrohling commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -119,13 +134,98 @@ public Builder withAuth(String authToken) {
+                String hash = CertificatePinner.pin(cert);
+                for (String pin: pins) {
+                  if (hash.equals(pin)) {
+                    return true;
```
Do I read this right, that the pin works for any cert in the chain, instead of "all" certs?

iori-yja replied:
Yes, pinning is to "ensures that there is an intersection between the public keys in the computed trust chain and the set of fingerprints associated with that domain".
https://wiki.mozilla.org/SecurityEngineering/Public_Key_Pinning

jpkrohling replied:
TIL, thanks!

jpkrohling commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -119,13 +134,98 @@ public Builder withAuth(String authToken) {
+                }
+              }
+              // TODO: For TOFU (trust on first use) usecase, print the chain
+              throw new CertificateException();
```
If there's anything wrong, it would be very hard to figure out *what* is wrong. I see the following possible error conditions:

* Empty chain (bad configuration from the user's part)
* No matches between the hostname and the subject CNs from the chain
* The check failed

It would be good to have an exception thrown with the appropriate message, even if it's a bit more code.

iori-yja replied:
`CertificateException` is thrown when the certificate couldn't verified _by the `TrustManager`_. If another (like default) manager might succeed the verification even after _this_ manager failed, the connection must be sane.

To follow this (strange) interface, I throw `CertificateException` whenever this manager fails to verify the connection.

But probably other exception could be thrown to fail in case of "The check failed" but CN matches including server's chain was empty.

iori-yja replied:
> But probably other exception could be thrown to fail in case of "The check failed" but CN matches including server's chain was empty.

This is good for trust-on-first-use use-case to print actual pins of the server, because calculating the pin by `openssl` command is a little bothering.

jpkrohling replied:
+1 for having different exceptions for the `check failed` case

iori-yja commented on jaeger-crossdock/https-proxy/proxy.template.conf:
```diff
@@ -0,0 +1,20 @@
+}
+
+server {
+    listen 14443 ssl;
```
I referred jaeger-collector's ports which range over 142xx (except zipkin's). This naming is 443 version of those ports -- but 8443 is also nice. Should I change it to 8443?

jpkrohling replied:
It's up to you. I personally would prefer 8443, but 14443 is also acceptable.

iori-yja commented on jaeger-crossdock/https-proxy/authority.crt:
```diff
@@ -0,0 +1,10 @@
+-----BEGIN CERTIFICATE-----
```
Noted in [gen.sh](https://github.com/jaegertracing/jaeger-client-java/pull/602/files#diff-36b2147f519397c9d0f88174ff9b306eR14)

(https://github.com/jaegertracing/jaeger-client-java/pull/602#discussion_r266480687)

pavolloffay commented:
@iori-yja is this ready from your side? @jpkrohling could you please check if your review has been fixed?

jpkrohling commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -119,13 +134,68 @@ public Builder withAuth(String authToken) {
       return this;
     }
 
+    public Builder withCertificatePinning(String[] sha256certs /* comma separated */) {
```
What is comma separated that the comment refers to?

iori-yja replied:
That _was_ comma separated but the argument itself is already processed to an array so it doesn't deal with the format. Just deleted, thank you to notice me.

jpkrohling commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -119,13 +134,68 @@ public Builder withAuth(String authToken) {
+        final URI uri = new URI(endpoint);
+        String hostname = hostname = uri.getHost();
+
+        if (uri.getScheme() != null && uri.getScheme().equals("https")) {
```
You can avoid the null check if you use this: 
```java
"https".equals(uri.getScheme())
```

iori-yja replied:
Thanks!

jpkrohling commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -119,13 +134,68 @@ public Builder withAuth(String authToken) {
+        clientBuilder.sslSocketFactory(sslContext.getSocketFactory(), (X509TrustManager) selfSignedServerTrustManager[0]);
+
+      } catch (java.security.NoSuchAlgorithmException e) {
+          // TLS is hardcoded abave. No occasion to come here.
```
s/abave/above

jpkrohling commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -119,13 +134,68 @@ public Builder withAuth(String authToken) {
+      } catch (java.security.NoSuchAlgorithmException e) {
+          // TLS is hardcoded abave. No occasion to come here.
+          throw new RuntimeException(e);
+      } catch (java.security.KeyManagementException e) {
```
Any reasons not to import the class, instead of referring it to its fully qualified name?

jpkrohling commented on jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java:
```diff
@@ -139,5 +209,60 @@ public Response intercept(Chain chain) throws IOException {
       };
     }
+
+    private static class SelfSignedTrustManager implements javax.net.ssl.X509TrustManager {
```
Any reasons not to import the class?

iori-yja commented:
Sorry, forgot to check; I will catch up this week

iori-yja commented:
> HTTPS URL succeeds when the server sends a self-signed cert and acceptSelfSigned() is used

`java-https` does this in crossdock tests.

> By default, HTTPS URL should not accept self-signed certs

This is the difficult part. How can I set "must-fail" test in crossdock?

> HTTP + acceptSelfSigned() is noop (or has no bad consequences)

`setTLSCertificatePinning()` test does this check, but the name may be misleading. I've changed it to `sendPlainHttpWithCertificatePinning()`.
